### ISSUE TITLE
Dependencies

### ISSUE THREAD

**Issue by ianpurton** [LouisGariepy/cornucopia#16]
Hi, I'm trying out cornucopia and so far I really like it.

I have to add cornucopia to my Cargo.toml, this then adds a lot of dependencies (40+) but we only use `cornucopia::GenericClient`?  

so 

```sql
--! example_query(id) {id, email, reset_password_selector?}*
SELECT 
    id, email, reset_password_selector
FROM 
    users
WHERE
    users.id < $1
```

generates 

```
pub mod types {
            
} 
 pub mod queries { pub mod user {

use cornucopia::GenericClient;
use tokio_postgres::Error;

#[derive(Debug, Clone, PartialEq)]
pub struct ExampleQuery {pub id : i32,pub email : String,pub reset_password_selector : Option<String>}
    pub async fn example_query<T: GenericClient>(client:&T, id : &i32) -> Result<Vec<super::super::queries::user::ExampleQuery>,Error> {let stmt = client.prepare("SELECT
id, email, reset_password_selector
FROM
users
WHERE
users.id < $1
").await?;
let res = client.query(&stmt, &[&id]).await?;

let return_value = res.iter().map(|res| { let return_value_0: i32 = res.get(0); let return_value_1: String = res.get(1); let return_value_2: Option<String> = res.get(2); super::super::queries::user::ExampleQuery { id : return_value_0,email : return_value_1,reset_password_selector : return_value_2 } }).collect::<Vec<super::super::queries::user::ExampleQuery>>(); Ok(return_value)}
} }
```

Could we make it so that use `cornucopia::GenericClient` isn't needed and then a project has less dependencies as we would not have cornucopia in the cargo.toml?

Perhaps take a Pool as a parameter?

**Comment by LouisGariepy** [comment]
Hi @ianpurton, this is a great suggestion. I originally thought `cornucopia`'s `GenericClient` would be very lightweight dependency since it relies pretty much entirely on `deadpool-postgres` and `tokio-postgres` (which are already needed to use cornucopia anyway), **but** it does also uses `async_trait` which might be where the extra dependencies are coming from.

Moving forward, I'd like to make it possible to specify a  `--connection-type` flag to the cli, with `generic` and `pool` being options. The `pool` option would achieve what you're suggesting.

I'm not sure how to handle transactions though. *If I were to introduce a `pool` option, to replace `GenericClient`s with simple pool `Client`s then your queries can no longer accept transactions (this is the whole purpose of the `GenericClient`). Is this something you're willing to give up? Or would you rather have `cornucopia` generate two versions of your queries, one for single statements, and one for transactions?*

This is really the crux of this issue for me, and it'd be really helpful to have a user's perspective on it. Thanks again for the feedback :)

**Comment by LouisGariepy** [comment]
Hi @ianpurton, this is a great suggestion, and I am open to support it. As a first step, I moved `GenericClient` to a separate crate. This should reduce the number dependencies quite a bit. This is reflected in the most recent patch (0.4.5), with clarifications in the readme and example.

 Could you please 
 1. Update the cli (`cargo install cornucopia`),
 2. Delete your generated queries (`cornucopia.rs` by default)
 3. Update your `Cargo.toml` to use `cornucopia_client` version 0.1.0, instead of `cornucopia`
 4. Regenerate your queries (`cornucopia generate` by default)
 
 `cornucopia_client` should be way more lightweight than `cornucopia` since it only has the necessary dependencies for the generic client. Do you still observe a large number of dependencies being pulled?
 
 If this is not resolved, I have further ideas that could solve this issue too. Let me know what you think, and we'll proceed from there!
 
 Thanks for using Cornucopia!

**Comment by ianpurton** [comment]
@LouisGariepy I followed all the steps and it works.

I have some more feedback.

When I use tools like https://github.com/hyperium/tonic the code generation is part of the build.rs and this generates the code into the target folder so it's not part of the project.

It then gets included into the source something like this in main.rs

```
pub mod statics {
    include!(concat!(env!("OUT_DIR"), "/cornucopia.rs"));
}
```

So somehow I will call cornucopia from build.rs only when the sql files change. Not sure how yet, but just to let you know.

For people who are happy to generate the code then check it into git, then it might be useful to do 2 things.

1. Call out to cargo fmt, so the code is formatted. (Assuming it is installed)
2. Add a comment at the top, something like 'This code is auto generated by cornucopia' so anyone reading the code can get an idea where it came form.

Thanks

**Comment by LouisGariepy** [comment]
@ianpurton. Thanks for the follow-up :smiley: .

__Regarding running cornucopia in `build.rs` when your sql files change:__
Right, that is something you can achieve right now, pretty easily too. Maybe what I can do is open a PR to add a new example using `build.rs` showcasing how to achieve what you're suggesting. https://github.com/LouisGariepy/cornucopia/issues/18.

__Regarding checking out cornucopia files into git:__
1. *Formatting the code*: this was already implemented, but in a limited way. I'll open an issue  to make the formatting options more powerful and clearer. https://github.com/LouisGariepy/cornucopia/issues/19.
2. *Top comment for generated code*: Good suggestion, will open an issue for that too. https://github.com/LouisGariepy/cornucopia/issues/20.

For now, I'll close this issue since the initial issue has been resolved (feel free to reopen if this is not the case!), but we can continue this discussion in one of the relevant issues if you have more questions or suggestions.

**Comment by LouisGariepy** [comment]
@ianpurton I just released version 0.5.0 which should address most of the feedback expressed here https://github.com/LouisGariepy/cornucopia/issues/16#issuecomment-1104837516. I'd be grateful if you could let me know if more work is needed in that regard. Feel free to open a new issue or a discussion on github. Thanks!

**Comment by ianpurton** [comment]
@LouisGariepy Hi, I've tested the build.rs and it's working fine.

I'm moving over a code base to cornucopia in the next few days, then I'll be able to give some better feedback. Thanks.
